### PR TITLE
Make the adaptive shift optimizer work correctly with OpenMP.

### PR DESCRIPTION
**Pull request by chrislzhao**
@ye-luo The new optimizer should work with OpenMP now. So please take a look.

### CODE REVIEW

**Comment by ye-luo** [comment]
I got an error to build the code
/sandbox/opt/optimizer-OMP/src/formic/utils/lmyengine/matrix_builder.cpp(248): error: namespace "std" has no member "cout"
    std::cout << "total weight is " << total_weight << std::endl;
         ^

compilation aborted for /sandbox/opt/optimizer-OMP/src/formic/utils/lmyengine/matrix_builder.cpp (code 2)
make[2]: *** [src/formic/utils/CMakeFiles/formic_utils.dir/lmyengine/matrix_builder.cpp.o] Error 2
make[1]: *** [src/formic/utils/CMakeFiles/formic_utils.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

**Comment by chrislzhao** [comment]
Hi Ye:
  That line was used for debugging and I commented it out.
Best,
Luning

**Comment by ye-luo** [comment]
I added a threaded test case. It shows the code is buggy.

      Start  50: short-H4-opt-adaptive-1-16
*** Error in `/sandbox/opt/optimizer-OMP/build_real/bin/qmcpack': double free or corruption (!prev): 0x000000000bb78e60 ***
 5/12 Test  #50: short-H4-opt-adaptive-1-16 .................***Failed   44.46 sec
      Start  51: short-H4-opt-adaptive-1-16-totenergy
 6/12 Test  #51: short-H4-opt-adaptive-1-16-totenergy .......***Failed    0.02 sec

**Comment by chrislzhao** [comment]
It should have been fixed by now.

**Comment by ye-luo** [comment]
/sandbox/opt/optimizer-OMP/src/formic/utils/lmyengine/matrix_builder.cpp(287): error: namespace "std" has no member "cout"
      std::cout << _hmat.print("%12.6f", "hamiltonian");
           ^

/sandbox/opt/optimizer-OMP/src/formic/utils/lmyengine/matrix_builder.cpp(290): error: namespace "std" has no member "cout"
      std::cout << _smat.print("%12.6f", "overlap");
           ^

compilation aborted for /sandbox/opt/optimizer-OMP/src/formic/utils/lmyengine/matrix_builder.cpp (code 2)
make[2]: *** [src/formic/utils/CMakeFiles/formic_utils.dir/lmyengine/matrix_builder.cpp.o] Error 2
make[1]: *** [src/formic/utils/CMakeFiles/formic_utils.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

**Comment by markdewing** [comment]
I see a similar error with double deletes.  It appears ReusableArray has methods for thread_lock/unlock.   Adding OMP locks there seems to fix the issue as well.   Maybe setting the matrix sizes initial also fixes it (no need to resize?).

Code looks something like this (not production ready - initialization of the lock is not correct):
<pre>
--- a/src/formic/utils/reusable_array.cpp
+++ b/src/formic/utils/reusable_array.cpp
@@ -22,6 +22,11 @@ template<class S> formic::ReusableArray<S>::ReusableArray(const size_t n)
   : m_length(n),
     m_array()
 {
+  if (!reusable_array_lock_initialized)
+  {
+    omp_init_lock(&reusable_array_lock);
+    reusable_array_lock_initialized = true;
+  }
 
   if ( n > 0 ) {
 
@@ -103,3 +108,6 @@ void formic::reusable_array_garbage_collect() {
   formic::ReusableArray<double>::garbage_collect();
   formic::ReusableArray<std::complex<double> >::garbage_collect();
 }
+
+template<typename T> omp_lock_t formic::ReusableArray<T>::reusable_array_lock;
+template<typename T>bool formic::ReusableArray<T>::reusable_array_lock_initialized = false;
diff --git a/src/formic/utils/reusable_array.h b/src/formic/utils/reusable_array.h
index 87f5070..db2a126 100644
--- a/src/formic/utils/reusable_array.h
+++ b/src/formic/utils/reusable_array.h
@@ -12,6 +12,7 @@
 #include <stack>
 
 #include <boost/shared_array.hpp>
+#include <omp.h>
 
 namespace formic {
 
@@ -42,10 +43,17 @@ namespace formic {
 
       // disable copy construction
       ReusableArray(const ReusableArray &);
+      static omp_lock_t reusable_array_lock;
+      static bool reusable_array_lock_initialized;
 
-      static void thread_lock() {}
 
-      static void thread_unlock() {}
+      static void thread_lock() {
+        omp_set_lock(&reusable_array_lock);
+      }
+
+      static void thread_unlock() {
+        omp_unset_lock(&reusable_array_lock);
+      }
 
     public:

</pre>

**Comment by chrislzhao** [comment]
I commented those two output lines out.

**Comment by prckent** [comment]
Small issue, perhaps a missing #include <iostream>? Given the issue, I don't understand how others have built this:

/home/pk7/projects/qmc/git_QMCPACK_luning/qmcpack/src/formic/utils/lmyengine/matrix_builder.cpp(287): error: namespace "std" has no member "cout"
      std::cout << _hmat.print("%12.6f", "hamiltonian");
           ^

/home/pk7/projects/qmc/git_QMCPACK_luning/qmcpack/src/formic/utils/lmyengine/matrix_builder.cpp(290): error: namespace "std" has no member "cout"
      std::cout << _smat.print("%12.6f", "overlap");
           ^

Configure line:
cmake -DCMAKE_C_COMPILER=mpiicc -DCMAKE_CXX_COMPILER=mpiicpc ../qmcpack/
using Intel2017 compilers & MPI. Setting -DBUILD_LMYENGINE_INTERFACE=0 correctly avoids this problem, but of course then the optimizer is not built.

**Comment by ye-luo** [comment]
The simple tests pass.
 3/12 Test  #48: short-H4-opt-adaptive-16-1 .................   Passed   99.47 sec
      Start  49: short-H4-opt-adaptive-16-1-totenergy
 4/12 Test  #49: short-H4-opt-adaptive-16-1-totenergy .......   Passed    0.02 sec
      Start  50: short-H4-opt-adaptive-1-16
 5/12 Test  #50: short-H4-opt-adaptive-1-16 .................   Passed  110.77 sec
      Start  51: short-H4-opt-adaptive-1-16-totenergy
 6/12 Test  #51: short-H4-opt-adaptive-1-16-totenergy .......   Passed    0.02 sec

**Comment by prckent** [comment]
Working here too (oxygen, Intel 2017):
      Start  48: short-H4-opt-adaptive-16-1
 3/12 Test  #48: short-H4-opt-adaptive-16-1 .................   Passed   56.95 sec
      Start  49: short-H4-opt-adaptive-16-1-totenergy
 4/12 Test  #49: short-H4-opt-adaptive-16-1-totenergy .......   Passed    0.02 sec
      Start  50: short-H4-opt-adaptive-1-16
 5/12 Test  #50: short-H4-opt-adaptive-1-16 .................   Passed   68.86 sec
      Start  51: short-H4-opt-adaptive-1-16-totenergy
 6/12 Test  #51: short-H4-opt-adaptive-1-16-totenergy .......   Passed    0.02 sec

**Comment by markdewing** [comment]
This doesn't build if OpenMP is not enabled - various omp_ functions are not found at link time.
To disable OpenMP - go to CMake/GNUCompilers.cmake and comment out SET(ENABLE_OPENMP 0) and the following two lines that add the openmp flags.

The solution in the rest of qmcpack is in src/Message/OpenMP.h, which defines dummy functions for omp_get_thread_num, etc. when ENABLE_OPENMP is not defined.
You could either include that file, or to keep the formic library independent, copy the contents to a similar file in the formic library.

**Comment by chrislzhao** [comment]
This should have been fixed by now.

**Comment by prckent** [comment]
@markdewing Can you confirm that this is OK to merge?

[ Continuous integration will be able to catch this in future. We should do a better job advertising that we aim for the code to build with or without either MPI or OpenMP. ]

File: src/formic/utils/openmp.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//
+// File created by: Jeongnim Kim, jeongnim.kim@gmail.com, University of Illinois at Urbana-Champaign
+//////////////////////////////////////////////////////////////////////////////////////
+    
+    
+
+
+
+
+#ifndef formic_OHMMS_OPENMP_H
```
Comment (markdewing) [inline src/formic/utils/openmp.h]: The guard name should be something like FORMIC_OPENMP_H
(to match the style, the guard name should be capitilized)

File: src/formic/utils/openmp.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+inline omp_int_t omp_get_max_threads()
+{
+  return 1;
+}
+inline omp_int_t omp_get_num_threads()
+{
+  return 1;
+}
+#endif
+#endif // OHMMS_COMMUNICATE_H
```
Comment (markdewing) [inline src/formic/utils/openmp.h]: The comment should match the guard name.

File: src/formic/utils/lmyengine/matrix_builder.cpp [hunk 3]
```diff
@@ -15,7 +15,8 @@
 #include<boost/format.hpp>
 #include<boost/shared_ptr.hpp>
 
-#include<omp.h>
+#include<formic/utils/openmp.h>
+//#include<omp.h>
```
Comment (markdewing) [inline src/formic/utils/lmyengine/matrix_builder.cpp]: Commented out include of omp.h should be removed

**Comment by chrislzhao** [comment]
I've made the change.

**Comment by prckent** [comment]
Thanks for the updates on this one.

I'll merge at 5.30pm EST unless any more comments are made.

**Comment by markdewing** [comment]
Some small items should be changed - noted in inline comments.
(//#include <omp.h> lines should be removed, and the include guard name in openmp.h should be changed)

**Comment by chrislzhao** [comment]
I have changed this.

**Comment by markdewing** [comment]
Thanks for the fixes.  I'm okay with merging.

**Pull request merged**